crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2845,14 +2846,218 @@ impl<'db> Type<'db> {
                 binding.into_outcome()
             }
 
-            // TODO annotated return type on `__new__` or metaclass `__call__`
-            // TODO check call vs signatures of `__new__` and/or `__init__`
-            Type::ClassLiteral(ClassLiteralType { .. }) => {
-                let signature = Signature::new(Parameters::gradual_form(), self.to_instance(db));
-                let binding = bind_call(db, arguments, &signature.into(), self);
+            // TODO: currently we can't use typeshed to infer calls using generic
+            // logic below due to a salsa query cycle. When this is fixed, we should
+            // this arm and aboid special casing `TypeVar` in the `ClassLiteral` arm.
+            Type::ClassLiteral(ClassLiteralType { class })
+                if class.is_known(db, KnownClass::TypeVar) =>
+            {
+                // TODO: signature depends on Python version, using 3.11 for now
+                // ```py
+                // class TypeVar:
+                //     elif sys.version_info >= (3, 11):
+                //        def __new__(
+                //            cls, name: str, *constraints: Any, bound: Any | None = None, covariant: bool = False,
+                //            contravariant: bool = False
+                //        ) -> Self: ...None: ...
+                // ```
+                #[salsa::tracked(return_ref)]
+                fn overloads<'db>(db: &'db dyn Db) -> CallableSignature<'db> {
```

Comment:
Can you explain what the reason is that `from_overloads` is in a salsa query? Is there a performance benefit or is it to avoid the salsa cycle?

Reply:
Thanks for explaining. I noticed when reviewing @dcreager's https://github.com/astral-sh/ruff/pull/16716 PR that he moved the query to a different method and I left a comment there. I think we can ignore this concern for this PR and come to a conclusion on https://github.com/astral-sh/ruff/pull/16716